### PR TITLE
[Fix][DSV4.1][SM90] Avoid materializing the full decode indexer slot map

### PR DESCRIPTION
**Pull request by Frank-whw**
## Motivation

This is a `main`-targeted resubmission of my earlier fix in #39979, submitted on September 17. That PR was closed when its target branch, `dsv4.1`, was deleted after #38798 merged. This version adapts the fix to the consolidated `fp4_indexer.py` on `main` and adds coverage for changing request mappings during CUDA graph replay and unallocated invisible slots.

Related: #40162 also resolves slots inside the kernel as part of a broader Hopper candidate-compact scoring change. This PR isolates the slot-map allocation fix, leaving the current candidate metadata, scoring, and top-k flow unchanged. Happy to coordinate the merge order or share these regression tests if maintainers prefer landing the broader change first.

At base `9784d5f979c9eaa5a6d9b31e596a7a73cb76dab8`, `_low_ratio_index_topk_sm90_decode` still materializes:

```python
slots = req_to_token[req[:, None], (j * ratio)[None, :]].to(torch.int64) // ratio
```

The allocation scales with the configured maximum compressed context width, not the actual visible length. At 1,536 verify rows and a ratio-1 width of 1,048,576, one int64 slot map alone is 12 GiB. The gather, conversion, division and masking also create temporaries. This can exhaust the remaining HBM during long-context DSpark startup/warmup. The row count depends on the warmup and parallelism configuration; `64 * 6` rows would be 3 GiB, not 12 GiB.

## Modifications

- Add a paged SM90 scoring entry point that resolves `req_to_token[req, logical_position * ratio] // ratio` inside each Triton tile.
- Resolve only the selected `[B, k]` slots after top-k, instead of gathering them from a full `[B, L]` slot map.
- Keep the existing materialized-slot entry point unchanged as the regression reference. Preserve FP4 decoding, BF16 rounding, candidate masks, sorting and output padding.
- Validate the paged lookup's mapping layout and bounds, including rejecting non-unit inner stride.
- Add SM90 regression tests registered for the existing `base-b-kernel-unit` / `1-gpu-large` CI configuration.

This removes the full slot-map allocation, not every context-proportional allocation. The `[B, L]` logits and candidate-mask temporaries remain. It does not include compact scoring, skip-empty-tile optimization, or a fused top-k change.

## Accuracy Tests

Fresh validation of the candidate sources on H200 (SM90), PyTorch 2.13.0+cu129, CUDA 12.9, Triton 3.7.1:

- All **5 registered test methods passed, with no skips**.
- **30 bitwise logits comparisons**: ratios 1/2, batches 1/7/64, widths 1/63/64/65/513. Includes repeated request IDs, padded mapping rows, zero visible length and tile/top-k boundaries; compares int32 bit patterns, including `-inf` padding.
- **12 CUDA graph replays**: both ratios; lengths 0/1/63/64/512/513. Lengths, request IDs and mapping contents change after capture, including lengths greater than the capture-time length.
- Unallocated (`-1`) invisible mapping entries, empty batch, zero-width outputs, and rejection of non-unit inner stride.
- **48 additional backend comparisons** using the actual method extracted from the base and candidate sources: plain/source/consumer/empty-consumer, ratios 1/2, batches 1/7/64, with/without raw indices. Final physical indices, raw indices and published masks match. Query and weight projections use identical fixed inputs; this is not a full-model test.
- At `[1536, 1048576]`, the isolated old/new component logits also match bitwise.
- `git diff --check`, source compilation, and all applicable pre-commit hooks pass. Static AST comparison confirms the pre-existing combined module is unchanged and the scoring math after slot resolution is identical.

The candidate kernel, its pure-Torch dependency and test sources were loaded into an isolated process in the existing GPU environment. This was **not** a fresh installation or a full-service startup A/B of the rebased branch. Full-model quality, end-to-end throughput and same-configuration 1M DSpark startup still need integration validation.

Run the registered tests in an environment installed from this branch:

```bash
python test/registered/kernels/ops/attention/test_sm90_fp4_indexer.py
```

## Speed Tests and Profiling

Fresh isolated measurements for slot lookup plus logits scoring, ratio=1. Inputs use finite FP4 scales and randomized mappings into a small physical pool, with repeated request IDs; these synthetic inputs do not represent a model's full KV working set.

Latency: CUDA events over graph replay, 3 warmups, median of 7 rounds of 5 replays. Memory: a separate eager call's peak allocated delta, including outputs and temporaries, excluding resident inputs, model weights and KV pools. It is not a graph-pool or whole-service memory measurement.

| Rows / max width / visible length | Main latency | Paged latency | Main peak delta | Paged peak delta |
| --- | ---: | ---: | ---: | ---: |
| 6 / 1M / 1K | 1.557 ms | 1.319 ms | 142 MiB | 24 MiB |
| 6 / 1M / 1M | 1.575 ms | 1.332 ms | 142 MiB | 24 MiB |
| 192 / 256K / 1K | 12.087 ms | 10.445 ms | 1,012 MiB | 192 MiB |

An additional isolated allocation check at **1,536 rows / 1M width / 1K visible** measured **32,272 MiB → 6,144 MiB**. No full model was resident. The remaining 6 GiB is the logits output; this is not a claim that the full 1M DSpark configuration now fits on every deployment.

## Checklist

- [x] All applicable pre-commit hooks and `git diff --check` pass.
- [x] Added a CI-registered SM90 regression test.
- [x] Provided fresh component correctness, memory and latency results, with integration-test limitations stated above.
- [x] No user-facing configuration or documentation changes are required.
- [x] Kept the patch scoped to the existing SM90 decode/verify slot-map allocation.

---
### CI States

Latest PR Test (Base): :x: [Run #35339214893](https://github.com/sgl-project/sglang/actions/runs/35339214893)
Latest PR Test (Extra): :x: [Run #35339214551](https://github.com/sgl-project/sglang/actions/runs/35339214551)
Latest PR Test (AMD ROCm 10): :x: [Run #35339214770](https://github.com/sgl-project/sglang/actions/runs/35339214770)

### CODE REVIEW

**Comment by songkuan-zheng** [comment]
Independent data point confirming this from the service side, on 8 x H200 SM90 / TP4 / DSpark k=5 / `--max-running-requests 8`.

The allocation being sized by configured maximum rather than visible length is visible from config alone: changing `--context-length` between 131072 and 1048576, at byte-identical prompts and fixed batch, moves per-request decode by 2.68x at bs=8 (1.50x at bs=1) — a per-row fixed-width cost, strictly linear in width (halving the configured width halves the saving: 10.91 : 5.52 ms). That quantifies the setup cost your patch removes at bs=1, where batch amortisation is absent.

Two things from our side that bear on the wording, if useful:

- The slot map is capacity-sized *by construction*, not incidentally: it inherits from `MAX_SEQ_LEN_FOR_CAPTURE = req_to_token.shape[1]` and the `page_table` width, both of which follow `--context-length`. So the memory blow-up you describe cannot be worked around from config today — the decode graph has no independent width (see the issue I just filed).
- Your 1536-verify-row / ratio-1 / 12 GiB figure is consistent with what we see at k=5 draft on a smaller row count: allocation scales with capacity, so a 1M-context deployment carries the 1M cost even when every live context is 2K.

Also +1 on isolating this from #40162 rather than folding it in — a memory-only fix is independently verifiable, and the two failure modes (HBM exhaustion during warmup vs steady-state decode cost) have different blast radii.
